File: Swift/PhaseJumpPro/Tests/PhaseJumpProTests/UpdatableTests.swift
import Foundation
import Testing
@testable import PhaseJumpPro

struct UpdatableTests {
    class Counter {
        var finishCount = 0
        var didFinish = false
        var timeValue = 0.0
    }
    
    @Test
    func isFinished() async throws {
        let counter = Counter()
        let sut = Updatable()
        
        sut.onFinishFunc = {
            counter.didFinish = true
        }
        sut.onUpdateFunc = { _ in
            .finish
        }
        
        #expect(false == sut.isFinished)
        sut.onUpdate(time: .init(delta: 1))
        #expect(sut.isFinished)
        #expect(counter.didFinish)
    }
    
    @Test
    func finish() async throws {
        let counter = Counter()
        let sut = Updatable()
        
        sut.onFinishFunc = {
            counter.didFinish = true
        }
        #expect(false == sut.isFinished)
        sut.finish()
        #expect(sut.isFinished)
        #expect(counter.didFinish)
    }
    
    @Test
    func update() async throws {
        let counter = Counter()
        let sut = Updatable()
        
        sut.onUpdateFunc = { time in
            counter.timeValue += time.delta
            return .continue
        }
        #expect(false == sut.isFinished)
        
        sut.onUpdate(time: .init(delta: 3))
        #expect(3 == counter.timeValue)
    }
    
    @Test
    func updateAfterfinish() async throws {
        let counter = Counter()
        let sut = Updatable()
        
        sut.onUpdateFunc = { time in
            counter.timeValue += time.delta
            return .continue
        }
        sut.finish()
        #expect(sut.isFinished)
        
        sut.onUpdate(time: .init(delta: 3))
        #expect(0 == counter.timeValue)
    }
}

File: PhaseJump/Xcode/PhaseJump/PhaseJumpTests/G Tests/Core/UpdatableTests.cpp
#include "gtest/gtest.h"
#include "Updatable.h"

using namespace PJ;
using namespace std;

namespace UpdatableTests {
}

using namespace UpdatableTests;

TEST(Updatable, TestIsFinished)
{
    bool didFinish{};
    
    Updatable sut;
    sut.onFinishFunc = [&](auto& updatable) {
        didFinish = true;
    };
    sut.onUpdateFunc = [&](auto& updatable, auto time) {
        return FinishType::Finish;
    };
    EXPECT_FALSE(sut.IsFinished());
    sut.OnUpdate({1});
    EXPECT_TRUE(sut.IsFinished());
    EXPECT_TRUE(didFinish);
}

TEST(Updatable, Finish)
{
    bool didFinish{};
    
    Updatable sut;
    sut.onFinishFunc = [&](auto& updatable) {
        didFinish = true;
    };
    EXPECT_FALSE(sut.IsFinished());
    sut.Finish();
    EXPECT_TRUE(sut.IsFinished());
    EXPECT_TRUE(didFinish);
}

TEST(Updatable, Update)
{
    float timeValue{};
    
    Updatable sut;
    sut.onUpdateFunc = [&](auto& updatable, auto time) {
        timeValue += time.delta;
        return FinishType::Continue;
    };
    EXPECT_FALSE(sut.IsFinished());
    
    sut.OnUpdate({3});
    EXPECT_EQ(3, timeValue);
}


TEST(Updatable, UpdateAfterFinish)
{
    float timeValue{};
    
    Updatable sut;
    sut.onUpdateFunc = [&](auto& updatable, auto time) {
        timeValue += time.delta;
        return FinishType::Continue;
    };
    sut.Finish();
    EXPECT_TRUE(sut.IsFinished());
    
    sut.OnUpdate({3});
    EXPECT_EQ(0, timeValue);
}
